dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/queryengine/SourcesToStartingPoints.scala

```diff
@@ -9,48 +9,172 @@ import io.shiftleft.semanticcpg.language.operatorextension.allAssignmentTypes
+          val (startingPoints, methodTasks) = calculateStartingPoints(sources, executorService)
+          val startingPointFromUsageInOtherClasses =
+            calculateStatingPointsWithUsageInOtherClasses(methodTasks, cpg, executorService)
+          (startingPoints ++ startingPointFromUsageInOtherClasses)
+            .sortBy(_.source.id)
+        })
+        .getOrElse(Nil)
     } catch {
       case e: RejectedExecutionException =>
         log.error("Unable to execute 'SourceTravsToStartingPoints` task", e); List()
     } finally {
-      fjp.shutdown()
+      executorService.shutdown()
     }
   }
 
+  /** This will process and identify the starting points except the usage in other classes. This run will identify the
+    * required tasks for calculating starting points with usage in other classes.
+    *
+    * @param sources
+    *   \- Sources list
+    * @param executorService
+    *   \- Shared executor service to process the task in parallel
+    * @return
+    *   List of StartingPointWithSource and List of tasks for calculating starting points with usage in other classes
+    */
+  private def calculateStartingPoints(
+    sources: List[StoredNode],
+    executorService: ExecutorService
+  ): (List[StartingPointWithSource], List[UsageInput]) = {
+    val allExceptUsageInOtherClasses       = SourceStartingPointResultAggregator(sources.size)
+    val allExceptUsageInOtherClassesThread = new Thread(allExceptUsageInOtherClasses)
+    allExceptUsageInOtherClassesThread.setName("All except usage in other classes result aggregator")
+    allExceptUsageInOtherClassesThread.start()
+    sources.foreach(src =>
+      executorService.submit(new SourceToStartingPoints(src, allExceptUsageInOtherClasses.resultQueue))
+    )
+    allExceptUsageInOtherClassesThread.join()
+    (allExceptUsageInOtherClasses.finalResult.toList, allExceptUsageInOtherClasses.methodTasks.toList)
+  }
+
+  /** This will calculate starting points by finding the usage in other classes.
+    *
+    * @param methodTasks
+    *   \- Inputs required for processing
+    * @param cpg
+    *   \- cpg to get list of methods
+    * @param executorService
+    *   \- Shared executor service to process the task in parallel
+    * @return
+    *   List of StartingPointWithSource
+    */
+  private def calculateStatingPointsWithUsageInOtherClasses(
+    methodTasks: List[UsageInput],
+    cpg: Cpg,
+    executorService: ExecutorService
+  ): List[StartingPointWithSource] = {
+    val methods                   = cpg.method.l
+    val usageInOtherClasses       = SourceStartingPointResultAggregator(methods.size)
+    val usageInOtherClassesThread = new Thread(usageInOtherClasses)
+    usageInOtherClassesThread.setName("Usage in other classes result aggregator")
+    usageInOtherClassesThread.start()
+    methods.foreach(m =>
+      executorService.submit(new SourceToStartingPointsInMethod(m, methodTasks, usageInOtherClasses.resultQueue))
+    )
+    usageInOtherClassesThread.join()
```

Comment:
Same question/remark about order of results.

Reply:
final sorting on the result based on the original order of sources is done on combined results. That's how I could verify the consistency in results. - https://github.com/joernio/joern/pull/4429/files#diff-b887bf35647ceef08727f3fa1753c2cdcb473e966b53d598bbdf1b2f186dd6d4R40